src/self_test.rs

```diff
@@ -0,0 +1,153 @@
+        shell: Shell,
+        command: String,
+        output: Output,
+    },
+    /// Failed to execute command
+    #[error("Failed to execute command `{command}`",
+        command = .command,
+    )]
+    Command {
+        shell: Shell,
+        command: String,
+        #[source]
+        error: std::io::Error,
+    },
+}
+
+#[cfg(feature = "diagnostics")]
+impl crate::diagnostics::ErrorDiagnostic for SelfTestError {
+    fn diagnostic(&self) -> String {
+        let static_str: &'static str = (self).into();
+        let context = match self {
+            Self::ShellFailed { shell, .. } => vec![shell.to_string()],
+            Self::Command { shell, .. } => vec![shell.to_string()],
+        };
+        return format!(
+            "{}({})",
+            static_str,
+            context
+                .iter()
+                .map(|v| format!("\"{v}\""))
+                .collect::<Vec<_>>()
+                .join(", ")
+        );
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum Shell {
+    Sh,
+    Bash,
+    Fish,
+    Zsh,
+}
+
+impl std::fmt::Display for Shell {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.executable())
+    }
+}
+
+impl Shell {
+    pub fn all() -> &'static [Shell] {
+        &[Shell::Sh, Shell::Bash, Shell::Fish, Shell::Zsh]
+    }
+    pub fn executable(&self) -> &'static str {
+        match &self {
+            Shell::Sh => "sh",
+            Shell::Bash => "bash",
+            Shell::Fish => "fish",
+            Shell::Zsh => "zsh",
+        }
+    }
+
+    #[tracing::instrument(skip_all)]
+    pub async fn self_test(&self) -> Result<(), SelfTestError> {
+        let executable = self.executable();
+        let mut command = match &self {
+            // On Mac, `bash -ic nix` won't work, but `bash -lc nix` will.
+            Shell::Sh | Shell::Bash => {
+                let mut command = Command::new(executable);
+                command.arg("-lc");
+                command
+            },
+            Shell::Zsh | Shell::Fish => {
+                let mut command = Command::new(executable);
+                command.arg("-ic");
+                command
+            },
+        };
+
+        #[cfg(all(target_os = "linux", target_arch = "x86_64"))]
+        const SYSTEM: &str = "x86_64-linux";
+        #[cfg(all(target_os = "linux", target_arch = "x86"))]
+        const SYSTEM: &str = "x86-linux";
+        #[cfg(all(target_os = "linux", target_arch = "aarch64"))]
+        const SYSTEM: &str = "aarch64-linux";
+        #[cfg(all(target_os = "macos", target_arch = "x86_64"))]
+        const SYSTEM: &str = "x86_64-darwin";
+        #[cfg(all(target_os = "macos", target_arch = "aarch64"))]
+        const SYSTEM: &str = "aarch64-darwin";
+
+        command.arg(format!(
+            r#"nix build --no-link --expr 'derivation {{ name = "self-test-{executable}"; system = "{SYSTEM}"; builder = "/bin/sh"; args = ["-c" "echo hello > \$out"]; }}'"#
```

Comment:
We could probably get around needing a system constant by using `builtins.currentSystem` along with `--impure` (I don't remember if `--impure` is necessary with `--expr`, but probably is), but I don't think it matters either way.

Reply:
@grahamc suggested we throw in a timestamp from Rust here, too.